Create enterprise-roll-out-checklist.rst (#3388)
mattermost/docs: source/getting-started/enterprise-roll-out-checklist.rst

Enterprise Roll Out Checklist
==============================

> This checklist is intended to serve as a guide to Enterprises who are rolling out Mattermost to thousands of users.  It includes preparing for the roll out, rolling out to users and reviewing the roll out.  

Reviewer: ```suggestion
This checklist is intended to serve as a guide to Enterprises who are rolling out Mattermost to thousands of users.  It includes preparing for the roll out, rolling out to users, and reviewing the roll out.  
```

1. Prepare for the Roll Out
-----------------------
Much of the preparation work is focused on ensuring the environment is deployed and secured prior to onboarding users. 

> a. Define the Roll Out Project

Reviewer 2: One consideration is using "1a.", "1b." etc to clarify where they belong to when quickly scanning through. But I'm totally 0/5 and we don't have a precedence for it. Non-blocking

````````````````````````````
- Define key stakeholders and project team members
 - Example project team members: Project Manager, Network Administrator, Database Administrator, Corporate Directory Administrator, Security & Compliance Administrator(s), User Support, User Champions, User Trainers
  
- Define use cases and requirements for teams, their workflows and their integrations
 - Resource: https://mattermost.com/blog/27-things-enterprises-can-learn-startups-increase-productivity/
 
- Define success criteria, goals and metrics to measure success

>  - Resource:https://docs.mattermost.com/getting-started/implementation_plan.html

Reviewer 3: ```suggestion
 - Resource: https://docs.mattermost.com/getting-started/implementation_plan.html
```

  
- Create a Project Charter to document goals, tasks, deliverables, and decisions 
 - Get buy-in from project team members and key stakeholders on the project charter 

b. Validate Essential Security and Compliance Requirements
````````````````````````````````````````````````````````
- Review Mattermost security features
 - Resource: https://docs.mattermost.com/overview/security.html#security-features
 
- Determine monitoring requirements

>  - Database, network, storage, log integrity, etc. 

Reviewer 3: ```suggestion
 - Database, network, storage, log integrity
```

 - Identify fields for log management tools (e.g. Splunk Enterprise event data)

- Determine environment access policies
 - Network access, physical access, group controlled access

- Determine encryption policies
 - Resource: https://docs.mattermost.com/administration/encryption.html
 - Resource: https://docs.mattermost.com/install/transport-encryption/config.html

- Determine System Administration access policies
 - Identify list of users or groups who need administrative access for Mattermost System Console, Command Line Tools and API privileges

- Define and configure authentication policies
 - Resource: https://docs.mattermost.com/overview/auth.html 

>  - Determine requirements for multi-factor authentication

Reviewer: ```suggestion
- Determine requirements for multi-factor authentication
```

 - Resource: https://docs.mattermost.com/deployment/auth.html

- Configure and test SSO or Corporate Directory integration (SAML or AD/LDAP)
 - Resource: https://docs.mattermost.com/deployment/sso-saml.html
 - Resource: https://docs.mattermost.com/deployment/sso-ldap.html

- Define your mobile usage policy
 - Resource: https://docs.mattermost.com/mobile/mobile-overview.html 
 - Resource: https://docs.mattermost.com/deployment/mobile-app-deployment.html

Create Development, Staging, and Production Environments
````````````````````````````````````````````````````````
- Finalize production environment design basing hardware on expected usage and requirements for high availability
 - Resource: https://docs.mattermost.com/overview/architecture.html
 - Resource: https://docs.mattermost.com/deployment/deployment.html 
 - Resource: https://docs.mattermost.com/deployment/scaling.html 
 - Resource: https://docs.mattermost.com/deployment/cluster.html

- Create development and staging environments 

>  - Recommend using to test early configurations for database, authentication, file storage, Elasticsearch, prior to setting up high availability and load balancing.  

Reviewer: ```suggestion
 - Recommend using to test early configurations for database, authentication, file storage, Elasticsearch, prior to setting up high availability and load balancing 
```

>  - Recommend configuring staging to be an identical replication of your production environment.

Reviewer: ```suggestion
 - Recommend configuring staging to be an identical replication of your production environment
```

- Create production environment
 - Install Mattermost
  - Install the number of nodes based on your high availability requirements outlined in your production environment design

>   - Recommendation: Use Kubernetes and the Mattermost Operator, with external supported external database and file storage solutions.  This will also provide blue/green deployment, rolling upgrades and canary builds

Reviewer: ```suggestion
  - Recommendation: Use Kubernetes and the Mattermost Operator, with external supported external database and file storage solutions. This will also provide blue/green deployment, rolling upgrades, and canary builds
```

   - Resource: https://docs.mattermost.com/install/install-kubernetes.html 

>  - Install and configure Database

Reviewer: ```suggestion
 - Install and configure database
```

  - Install the number of read and search replicas based on your high availability requirements outlined in your production environment design

>    - Resources: https://docs.mattermost.com/overview/architecture.html#database-with-vips 

Reviewer 3: ```suggestion
   - Resource: https://docs.mattermost.com/overview/architecture.html#database-with-vips 
```

  - (Optional) Set up configuration management via the database instead of a config file for high available environments
   - Resource: https://docs.mattermost.com/administration/config-in-database.html
 - Install and configure File Storage
  - Resource: https://docs.mattermost.com/deployment/deployment.html#file-store

>  - Install and configure proxy or load balancers.  

Reviewer 3: ```suggestion
 - Install and configure proxy or load balancers
```

>   - Note: If running Kubernetes and the Mattermost Operator, proxies will be automatically created for you. 

Reviewer: ```suggestion
  - Note: If running Kubernetes and the Mattermost Operator, proxies will be created automatically. 
```

  - Add SSL Cert

>    - Resources: https://docs.mattermost.com/deployment/ssl-client-certificate.html

Reviewer: ```suggestion
   - Resource: https://docs.mattermost.com/deployment/ssl-client-certificate.html
```

   - Resource: https://docs.mattermost.com/deployment/cluster.html#proxy-server-configuration 
  - (Optional) Set up certificate-based authentication (CBA) for user or device-based authentication with a digital certificate
   - Resource: https://docs.mattermost.com/deployment/certificate-based-authentication.html
 - Configure SMTP for email notifications
  - Resource: https://docs.mattermost.com/install/smtp-email-setup.html
 - Set up Elasticsearch (highly recommended if your organization anticipates over 2 million posts)
  - Resource: https://docs.mattermost.com/deployment/elasticsearch.html

- Document network configuration
 - Example: https://docs.mattermost.com/overview/architecture.html#reference-architectures  

d. Configure and Customize Your Mattermost Site
````````````````````````````````````````````
- Login to Mattermost and access the System Console to connect your environment to Mattermost
 - Resource: https://docs.mattermost.com/administration/config-settings.html#environment
 - Upload your valid Enterprise License under Edition and License
 - Ensure site URL is set appropriately for your production, dev and staging environments

>  - Add your database configuration to Environment > Database 

Reviewer: ```suggestion
 - Add your database configuration to **Environment > Database**
```

>  - Add your Elasticsearch configuration to Environment > Elasticsearch 

Reviewer: ```suggestion
 - Add your Elasticsearch configuration to **Environment > Elasticsearch**
```

>  - Add your files storage configuration to Environment > File Storage 

Reviewer: ```suggestion
 - Add your file storage system configuration to **Environment > File Storage** 
```

>  - Add your proxy configuration to Environment > Image Proxy 

Reviewer: ```suggestion
 - Add your proxy configuration to **Environment > Image Proxy** 
```

>  - Add your SMTP configuration to Environment > SMTP

Reviewer: ```suggestion
 - Add your SMTP configuration to **Environment > SMTP**
```

>  - Enable Push notifications on Environment > Push Notification Server 

Reviewer: ```suggestion
 - Enable Push Notifications by adding your server in **Environment > Push Notification Server**
```

>  - Add your cluster configuration to Environment > High Availability

Reviewer: ```suggestion
 - Add your cluster configuration to **Environment > High Availability**
```

 - Resource: https://docs.mattermost.com/administration/config-settings.html#site-configuration

- Set site access policies including permissions for roles and guest access
 - Permissions Resource: https://docs.mattermost.com/deployment/advanced-permissions.html
 - Guest Access Resource: https://docs.mattermost.com/deployment/guest-accounts.html

e. Test Production Performance and Redundancy
``````````````````````````````````````````
- Define and test disaster recovery policy and processes
 - Resource: https://docs.mattermost.com/install/install-kubernetes.html#using-mattermost-operator-functionality
 - Resource: https://docs.mattermost.com/deployment/cluster.html#upgrade-guide 

- Performance test production environment
 - Load test production environment to verify that it will handle anticipated user load
  - Resource: https://github.com/mattermost/mattermost-load-test 
 - Set up Prometheus and Grafana to monitor performance
  - Resource: https://docs.mattermost.com/deployment/metrics.html 
 - Set up alerts in Grafana
  - Resource: https://docs.mattermost.com/administration/performance-alerting-guide.html 

2. Roll Out Mattermost
-----------------------

> Now that you have the an environment in place, we recommend working through the following items in an iterative process.  You may need to cycle through each of these topics multiple times to make adjustments to suit your organization as you on-board groups of users. 

Reviewer: ```suggestion
Now that you have an environment in place, we recommend working through the following items in an iterative process. You may need to cycle through each of these topics multiple times to make adjustments to suit your organization as you onboard groups of users. 
```

Define Your Team and Channel Strategy
`````````````````````````````````````

> - Determine and create team structure for your instance

Reviewer 2: ```suggestion
- Determine and create team structure for your environment
```

(Since we use the term `environment` elsewhere)

>  - Recommendation: Start with fewer teams in your early roll-out. 

Reviewer: ```suggestion
 - Recommendation: Start with fewer teams in your early roll-out
```

>  - Recommendation: Start with fewer teams in your early roll-out

Reviewer: ```suggestion
 - Recommendation: Start with fewer teams in your early roll out
```

 - Resource: https://docs.mattermost.com/help/getting-started/organizing.html 

> - Determine and create key channels to support your users. Town Square and Off-Topic are built in channels in every team. 

Reviewer: ```suggestion
- Determine and create key channels to support your users. Town Square and Off-Topic are built in channels in every team
```

>  - Recommendation: Add a “Support” channel for your users to escalate questions. 

Reviewer: ```suggestion
 - Recommendation: Add a “Support” channel for your users to escalate questions 
```

- (Optional) Migrate messages and channels from legacy systems 
 - Resource: https://docs.mattermost.com/administration/migrating.html

Enable Key Integrations
```````````````````````

> - Build list of key integrations and tools used by your teams. 

Reviewer: ```suggestion
- Build list of key integrations and tools used by your teams
```

 - Resource: https://docs.mattermost.com/guides/administrator.html#mattermost-integrations

> - Define use cases and requirements for plugins, bots, webhooks, plugins, slash commands 

Reviewer: ```suggestion
- Define use cases and requirements for plugins, bots, webhooks, slash commands 
```

 - Resource: https://docs.mattermost.com/guides/integration.html

- Set up key integrations (or migrate from POC environments)
 - Resource: https://integrations.mattermost.com/

- Understand Mattermost API capabilities
 - Resource: https://api.mattermost.com/

Prepare for User Onboarding
````````````````````````````
- Onboard champion users 

- Onboard trainers and support team
- Create a training plan

>  - Resource:  https://academy.mattermost.com/

Reviewer: ```suggestion
 - Resource: https://academy.mattermost.com/
```

>  - Resource: End User Guide: https://docs.mattermost.com/guides/user.html  

Reviewer 3: ```suggestion
 - Resource: End User Guide https://docs.mattermost.com/guides/user.html  
```

- Define user escalation and support processes

>  - Ensure you have set the site’s support url to your own support team under System Console > Site Configuration > Customization

Reviewer: ```suggestion
 - Ensure you have set the site’s support URL to your own support team under **System Console > Site Configuration > Customization**
```

> - Notify users in advance of rollout

Reviewer: ```suggestion
- Notify users in advance of roll out
```

>  - Sample Email: https://docs.mattermost.com/getting-started/welcome_email.html

Reviewer: ```suggestion
 - Sample email: https://docs.mattermost.com/getting-started/welcome_email.html
```

d. Deploy Client Apps
```````````````````
- Roll out Desktop App 
 - Resource: https://docs.mattermost.com/install/desktop.html
 - Resource: https://docs.mattermost.com/deployment/desktop-app-deployment.html
 - (Optional) Use the MSI installer to install on Windows machines
  - Resource: https://docs.mattermost.com/install/desktop-msi-gpo.html

- Roll out Mobile App
 - Resource: https://docs.mattermost.com/deployment/mobile-app-deployment.html
 - (Optional) Use an EMM provider
  - Resource: https://docs.mattermost.com/mobile/mobile-overview.html#use-an-emm-provider-with-managed-app-configuration 

e. Roll Out to Groups of Users
````````````````````````````
- Provision user accounts
 - Resource: https://docs.mattermost.com/administration/user-provisioning.html 

- (Optional) Bulk Load users
 - Resource: https://docs.mattermost.com/deployment/bulk-loading.html 

- Onboard users to teams and channels
 - Recommendation: Use LDAP Group Sync to automate this process
  - Resource: https://docs.mattermost.com/deployment/ldap-group-sync.html

> - Implement your training plan to end-users on how to use Mattermost 

Reviewer: ```suggestion
- Implement your training plan to end users on how to use Mattermost 
```

 - Train on using Mattermost

>  - Train on how to use Integrations

Reviewer: ```suggestion
 - Train on how to use integrations
```

f. Drive Adoption
```````````````
- Incrementally roll out to additional user groups
 - See “Roll out to groups of users”

> - Manage support requests and product requests from your end-users

Reviewer: ```suggestion
- Manage support requests and product requests from your end users
```

>  - See “Roll out to groups of users”

Reviewer: ```suggestion
 - See “Roll Out to Groups of Users”
```
It is possible to add a reference to this. It would be similar to "html/getting-started/enterprise-roll-out-checklist.html#e-roll-out-to-groups-of-users" and we can add that once it's published.

 - Resource: https://mattermost.com/support/ 
 - See process below for escalating to Mattermost

- Enable additional integrations and plugins to support user workflows
 - Resource: https://integrations.mattermost.com/

- Understand management tools available to support users
 - Command Line Tools Resource: https://docs.mattermost.com/administration/command-line-tools.html
 - Database Scripts Resource: https://docs.mattermost.com/administration/scripts.html 

3. Review the Roll Out 
-----------------------
We recommend that you review your roll out on a cadence that matches your iterative approach to rolling out to users.  Below are some areas to consider.  

> a. Review Project Charter success metrics

Reviewer: ```suggestion
a. Review Project Charter Success Metrics
```

``````````````````````````````
- Perform end-user surveys and measure satisfaction

>  - Optional Resource within Mattermost: https://integrations.mattermost.com/matterpoll/

Reviewer: ```suggestion
 - Optional resource within Mattermost: https://integrations.mattermost.com/matterpoll/
```

> - Verify use-case fulfillment from original requirements gathering

Reviewer: ```suggestion
- Verify use case fulfillment from original requirements gathering
```

- Measure your response time and resolution rate for user support issues

- Identify usage gaps and address or create a plan for addressing

b. Review and Analyze Usage
`````````````````````````
- Review Project Charter success metrics - identify usage gaps and address or create a plan for addressing

- Monitor site and team statistics 
 - Resource: https://docs.mattermost.com/administration/statistics.html 

>  - Review: Total Posts, total teams, total channels, total group chats, total direct chats, top channels, top teams

Reviewer: ```suggestion
 - Review: Total posts, total teams, total channels, total group chats, total direct chats, top channels, top teams
```

- Analyze usage by lines of business and peak usage times
 - Resources: https://docs.mattermost.com/administration/scripts.html

Analyze System Performance
````````````````````````````
- Monitor trends in CPU/memory usage

> - Review trends in Database connections

Reviewer: ```suggestion
- Review trends in database connections
```

- Review trends in Go routines 

- Review trends in concurrent sessions 

d. Harden Security
```````````````````
- Harden security controls around web, desktop and mobile security

- Harden configuration management 

- Harden network security
 - Identify additional tests and scans
 - (Optional) Enable Compliance Reporting
  - Resource: https://docs.mattermost.com/administration/compliance-export.html

e. Perform Maintenance Requests
````````````````````````````

> - Monitor for security updates (or sign up for email updates): 

Reviewer 3: ```suggestion
- Monitor for security updates (or sign up for email updates)
```

 - Resource: https://mattermost.com/security-updates/

- Perform first upgrade
 - Resource: https://docs.mattermost.com/administration/upgrade.html

> - Determine upgrade schedule based on Mattermost release schedules and life cycle

Reviewer: ```suggestion
- Determine upgrade schedule based on Mattermost release schedules and lifecycle
```

 - Resource: https://docs.mattermost.com/deployment/scaling.html 
 - Resource: https://docs.mattermost.com/deployment/cluster.html

- Create development and staging environments 
 - Recommend using to test early configurations for database, authentication, file storage, Elasticsearch, prior to setting up high availability and load balancing.  
 - Recommend configuring staging to be an identical replication of your production environment.

- Create production environment
 - Install Mattermost
  - Install the number of nodes based on your high availability requirements outlined in your production environment design
  - Recommendation: Use Kubernetes and the Mattermost Operator, with external supported external database and file storage solutions.  This will also provide blue/green deployment, rolling upgrades and canary builds
   - Resource: https://docs.mattermost.com/install/install-kubernetes.html 
 - Install and configure Database
  - Install the number of read and search replicas based on your high availability requirements outlined in your production environment design
   - Resources: https://docs.mattermost.com/overview/architecture.html#database-with-vips 

>   - (Optional)Set up configuration management via the database instead of a config file for high available environments

Reviewer 3: ```suggestion
  - (Optional) Set up configuration management via the database instead of a config file for high available environments
```
